samples/subsys/ipc/openamp/CMakeLists.txt

```diff
@@ -41,5 +41,6 @@ ExternalProject_Add(
   # NB: Do we need to pass on more CMake variables?
   BUILD_ALWAYS True
 )
+add_dependencies(app openamp_remote)
```

Comment:
This is not correct.

Building `app` is not in anyway dependent on the `openamp_remote` project.
I can remove all the external project code, and Zephyr will build and link just fine.

So this should not really be handled this way.

Reply:
Right but WHAT?
I could directly reference the "run" and "debugserver" targets here but that seems a bit clunky.
I guess the only handled emulator for these samples is QEMU right now anyway.

I have now spent many hours looking at the build system (at cmake level and generated GNU makefiles) and I still don't understand why flash works.  Best I can tell it has an implicit dependency on "all" potentially through a recursive call to cmake or make.

Reply:
> Right but WHAT? I could directly reference the "run" and "debugserver" targets here but that seems a bit clunky. I guess the only handled emulator for these samples is QEMU right now anyway.
> 
> I have now spent many hours looking at the build system (at cmake level and generated GNU makefiles) and I still don't understand why flash works. Best I can tell it has an implicit dependency on "all" potentially through a recursive call to cmake or make.

Because `flash` issues a complete build, and the `openamp_remote` target is part of `ALL`.
https://github.com/zephyrproject-rtos/zephyr/blob/5e07fabe0ccb6f8504934222b21a54c7861562d7/scripts/west_commands/flash.py#L28-L29
https://github.com/zephyrproject-rtos/zephyr/blob/5e07fabe0ccb6f8504934222b21a54c7861562d7/scripts/west_commands/run_common.py#L160-L161

Whereas, the qemu runner only runs that specific target, and in order to build that specific target, then the dependencies of that target is built.
Which in this case is specified here:
https://github.com/zephyrproject-rtos/zephyr/blob/5e07fabe0ccb6f8504934222b21a54c7861562d7/cmake/emu/qemu.cmake#L373
that resolve to:
https://github.com/zephyrproject-rtos/zephyr/blob/5e07fabe0ccb6f8504934222b21a54c7861562d7/CMakeLists.txt#L1347

so the right solution is in fact that the runner should have a dependency to `openamp_remote`.
Exactly because this is the real dependency.

Question is, should we adjust `cmake/emu/qemu.cmake` and friends to cleanly specify such dependencies without needing to loop all possible targets outside `qemu.cmake`.
Or should we instead do this work after #40555 has completed.

If you adjust the commit message, and describe that this is a sub-optimal approach and why it's considered as such (hint: having a dependency from app --> openamp_remote is not correct)
Then we can take this in until #40555 is complete.

No reason to do major re-work in this area for now.

Reply:
@tejlmand Commit message revised.